pkg/dynamicaccess/history.go

```diff
@@ -1,36 +1,158 @@
 package dynamicaccess
 
 import (
-	"github.com/ethereum/go-ethereum/common"
-	"github.com/ethersphere/bee/v2/pkg/kvs"
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"strconv"
+	"time"
+
+	"github.com/ethersphere/bee/v2/pkg/file"
+	"github.com/ethersphere/bee/v2/pkg/manifest"
+	"github.com/ethersphere/bee/v2/pkg/manifest/mantaray"
+	"github.com/ethersphere/bee/v2/pkg/swarm"
 )
 
 type History interface {
-	Add(timestamp int64, kvs kvs.KeyValueStore) error
-	Get(timestamp int64) (kvs.KeyValueStore, error)
-	Lookup(at int64) (kvs.KeyValueStore, error)
+	Add(ctx context.Context, actRef swarm.Address, timestamp *int64) error
+	Lookup(ctx context.Context, timestamp int64, ls file.LoadSaver) (swarm.Address, error)
+	Store(ctx context.Context) (swarm.Address, error)
 }
 
 var _ History = (*history)(nil)
 
+var ErrEndIteration = errors.New("end iteration")
+
 type history struct {
-	history map[int64]*kvs.KeyValueStore
+	manifest *manifest.MantarayManifest
+}
+
+func NewHistory(ls file.LoadSaver, ref *swarm.Address) (*history, error) {
```

Comment:
in kvs, and grantee the validity of a ref is checked against ZeroAddress and EmptyAddress. I think we should agree on a pattern and use that everywhere

Reply:
Ref also can be an invalid address as well. Do we really wanna do a full validation check in a single function? I am not sure what is the best approach in GO. In JS it's usually the task of a middleware.

Reply:
I switched to ZeroAddress